### PR TITLE
Revisit encoding/decoding implementation

### LINKED ISSUE #590: Cookie name with slash
**Describe the bug**
Cookie name with slash `/` character is escaped to `%2F`
Eg. `user/123=foo` becomes `user%2F123=foo`

I believe the forward slash character is allowed in the cookie key.

It's unclear why you're passing `key` here as the converter expects a single parameter:
`value = converter.write(value, key)`

And on the next line:
https://github.com/js-cookie/js-cookie/blob/b1d83a073525e00f87d8b4effa4804b8cb741f12/src/api.mjs#L21

Perhaps it would be better to have:
`key = converter.write(key)`

**Expected behavior**
Cookie name preserves the original forward slash character.

### PR DESCRIPTION
**Pull request by carhartl**
Our RFC 6265 based cookie name + value encoding was based on the
requirements for server implementers, and as such too strict.

We need to look at RFC 6265 section 5.2, requirements for user agents.

But even the algorithm described in section 5.2 isn't that relevant
for us, in that we had to follow the steps when writing a cookie with
this library. This is left to the browser vendors who have to implement
the setter functionality behind `document.cookie`. All we have to do is
avoid surprises when said algorithm is being followed while
`document.cookie = ...` is being executed with the cookie string we're
constructing.

It means we have to encode ";" and "=" in the cookie name, and ";" in
the cookie value.

Follow-up to discussion in #595.

Closes #590, #595.

~Open question: should we be liberal in what we accept/read, that is apply `decodeURIComponent()` to the cookie value, in case we get a server-side written cookie that adheres to RFC 6265? Since we will use percent encoding for ";" in the value we could support that without a lot of effort. An advantage might be that this way we'd be reading v2 cookies without a problem, making the transition from v2 -> v3 not require a cookie migration. Or would we leave that to the user who would have to implement a converter for a complete percent decoding?~

The upside is that we're again well-below 700kb for the gzipped size of the library.

PS: I would probably also want to rename the "rfc6265" converter to something else now as part of this... basically they're semicolon converters now.

### CODE REVIEW

**Comment by carhartl** [comment]
>Should we be liberal in what we accept/read, that is apply decodeURIComponent() to the cookie value when reading, in case we get a server-side written cookie that adheres to RFC 6265?

Wrong! Same mistake again. This is making the assumption that servers adhering to RFC 6265 would use percent-encoding to encode the forbidden characters mentioned in the spec, but the spec doesn't say anything about how those characters should be encoded. It mentions Base64 in fact:

>servers that wish to store arbitrary data in a cookie-value SHOULD encode that data, for example, using Base64

**Comment by carhartl** [comment]
So.. we are going to apply the minimal encoding/decoding, for everything else we are providing converters.

**Comment by carhartl** [comment]
I have no idea why the pr check fails in Travis, while the one for push passes...

**Comment by FagnerMartinsBrack** [comment]
Need to verify if [SERVER_SIDE.md](https://github.com/js-cookie/js-cookie/blob/d727ba6ceffb101e300a38b4eb06bcca4ac571d1/SERVER_SIDE.md) fixes still hold true

File: README.md [hunk 1]
```diff
@@ -166,12 +165,9 @@ _Note: The `.noConflict` method is not necessary when using AMD or CommonJS, thu
...
-The only character in cookie-name or cookie-value that is allowed and still encoded is the percent `%` character, it is escaped in order to interpret percent input as literal.  
+Special characters that are not permitted in the cookie name (";" and "=") or cookie value (";") are encoded with their UTF-8 Hex equivalent using [percent-encoding](http://en.wikipedia.org/wiki/Percent-encoding).
 Please note that the default encoding/decoding strategy is meant to be interoperable [only between cookies that are read/written by js-cookie](https://github.com/js-cookie/js-cookie/pull/200#discussion_r63270778). To override the default encoding/decoding strategy you need to use a [converter](#converters).
 
-_Note: According to [RFC 6265](https://tools.ietf.org/html/rfc6265#section-6.1), your cookies may get deleted if they are too big or there are too many cookies in the same domain, [more details here](https://github.com/js-cookie/js-cookie/wiki/Frequently-Asked-Questions#why-are-my-cookies-being-deleted)._
```
Comment (FagnerMartinsBrack) [inline README.md]: Why removing this?

File: Gruntfile.js [hunk 2]
```diff
@@ -1,33 +1,8 @@
...
-}
-
 const config = {
   qunit: {
     all: {
       options: {
-        urls: [
-          'http://127.0.0.1:9998/',
-          'http://127.0.0.1:9998/module.html',
-          'http://127.0.0.1:9998/encoding.html?integration_baseurl=http://127.0.0.1:9998/'
```
Comment (FagnerMartinsBrack) [inline Gruntfile.js]: Need also to remove [this part](https://github.com/js-cookie/js-cookie/blob/d727ba6ceffb101e300a38b4eb06bcca4ac571d1/CONTRIBUTING.md#integration-with-server-side) from the README which documents the `?integration_baseurl`.

**Comment by FagnerMartinsBrack** [comment]
> The upside is that we're again well-below 700kb for the gzipped size of the library.

Can change the README to <700kb gzipped then! :D

File: README.md (hunk 1 above)
Comment (carhartl) [reply README.md]: Still true, but imo in the wrong place, it doesn't really have anything to do with encoding.

Comment (carhartl) [reply README.md]: (And after all, there's an entry in the FAQ...)

**Comment by carhartl** [comment]
>Need to verify if SERVER_SIDE.md fixes still hold true

I looked at these and I think so.. or they've been wrong even before these changes here: for instance what confused me was that the docs suggest to escape parens for Tomcat:

```javascript
var TomcatCookies = Cookies.withConverter({
  write: function (value) {
    return (
      Cookies.converter
        .write(value)
        // Encode the parens that are interpreted incorrectly by Tomcat
        .replace(/[()]/g, escape)
    )
  },
  read: Cookies.converter.read
})
```

but the default converter was percent-encoding `(` and `)` anyway.

File: README.md (hunk 1 above)
Comment (carhartl) [reply README.md]: Ok to resolve?

Comment (FagnerMartinsBrack) [reply README.md]: Sure

**Pull request merged**